torch/_utils.py

```diff
@@ -127,19 +134,32 @@ def _unflatten_tensors(flat, tensors):
 
 
 def _take_tensors(tensors, size_limit):
-    """Groups tensors into lists of up to size_limit bytes"""
+    """Groups tensors into chunks. This generator yeilds a chunks at each call,
+    containing tensors up to certain byte limit in total size.
+
+    Args:
+        tensors (list): The list of tensors to be separated into chunks.
+        size_limit (int): The limit of each chunk in bytes.
+    """
     buf = []
-    size = 0
+    dense_size = 0
     last_type = type(tensors[0]) if len(tensors) > 0 else None
+    last_is_sparse = False
     for tensor in tensors:
         t = type(tensor)
-        param_size = tensor.numel() * tensor.element_size()
-        if t is not last_type or (size + param_size > size_limit and size > 0):
+        is_sparse = tensor.is_sparse
+        if not is_sparse:
+            param_size = tensor.numel() * tensor.element_size()
+        if last_is_sparse or \
+            t is not last_type or \
+            (not is_sparse and dense_size + param_size > size_limit and dense_size > 0):
             yield buf
             last_type = t
-            size = 0
+            dense_size = 0
             buf = []
         buf.append(tensor)
-        size += param_size
+        last_is_sparse = is_sparse
+        if not is_sparse:
+            dense_size += param_size
```

Comment:
Wouldn't be enough to just do this at the top of the loop:
```python
if tensor.is_sparse:
    yield tensor
    continue
```
So if you see a sparse tensor you just return it, and don't try to break dense groups into smaller subgroups

Reply:
Good point!

Reply:
Well we can make it maintain the order within dense/sparse classes :blush: